### PR TITLE
[ticket/13685] Allow modifying the query for mysql native fulltext search

### PR DESCRIPTION
**Pull request by brunoais**
https://tracker.phpbb.com/browse/PHPBB3-13685

### CODE REVIEW

File: phpBB/phpbb/search/fulltext_mysql.php [hunk 1]
```diff
@@ -447,6 +447,52 @@ public function keyword_search($type, $fields, $terms, $sort_by_sql, $sort_key,
...
+			'sql_sort_join',
+			'sort_days',
+			'sql_match',
+			'sql_match_where',
+			'sort_by_sql',
+			'sort_key',
+			'sort_dir',
+			'sql_sort',
+			'start',
+			);
```
Comment (nickvergessen) [inline phpBB/phpbb/search/fulltext_mysql.php]: -1 tab

**Comment by brunoais** [comment]
Restarting build....

File: phpBB/phpbb/search/fulltext_mysql.php [hunk 2]
```diff
@@ -447,6 +455,55 @@ public function keyword_search($type, $fields, $terms, $sort_by_sql, $sort_key,
...
+		*									Used in conjunction with sql_sort_table
+		* @var	int		sort_days			Time, in days, of the oldest possible post to list
+		* @var	string	sql_match			Which columns to do the search on.
+		* @var	string	sql_match_where		Extra conditions to use to properly filter the matching process
+		* @var	string	sort_by_sql			The possible predefined sort types
+		* @var	string	sort_key			The sort type used from the possible sort types
+		* @var	string	sort_dir			"a" for ASC or "d" dor DESC for the sort order used
+		* @var	string	sql_sort			The result SQL when processing sort_by_sql + sort_key + sort_dir
+		* @var	int		start				How many posts to skip in the search results (used for pagination)
+		* @since 3.1.4-RC1
```
Comment (marc1706) [inline phpBB/phpbb/search/fulltext_mysql.php]: 3.1.5-RC1

File: phpBB/includes/acp/acp_main.php [hunk 3]
```diff
@@ -632,7 +632,7 @@ function main($id, $mode)
 		{
 			$error = false;
 			$search_type = $config['search_type'];
-			$search = new $search_type($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $user);
+			$search = new $search_type($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $phpbb_dispatcher, $user);
```
Comment (nickvergessen) [inline phpBB/includes/acp/acp_main.php]: This will break all search engines that exist?

File: phpBB/phpbb/search/fulltext_mysql.php [hunk 4]
```diff
@@ -78,12 +84,14 @@ class fulltext_mysql extends \phpbb\search\base
 	 * @param \phpbb\auth\auth $auth Auth object
 	 * @param \phpbb\config\config $config Config object
 	 * @param \phpbb\db\driver\driver_interface Database object
+	 * @param \phpbb\event\dispatcher_interface	$phpbb_dispatcher	Event dispatcher object
 	 * @param \phpbb\user $user User object
 	 */
-	public function __construct(&$error, $phpbb_root_path, $phpEx, $auth, $config, $db, $user)
+	public function __construct(&$error, $phpbb_root_path, $phpEx, $auth, $config, $db, $phpbb_dispatcher, $user)
```
Comment (nickvergessen) [inline phpBB/phpbb/search/fulltext_mysql.php]: Please move the new entry to the end, so it doesn't hurt if another search engine is missing it.

Comment (nickvergessen) [inline phpBB/phpbb/search/fulltext_mysql.php]: or does @Nicofuma have a good idea how to fix this? Maybe a factory, or make them services or whatever?

Comment (brunoais) [inline phpBB/phpbb/search/fulltext_mysql.php]: Won't it be hurting because there's too many parameters?

Comment (brunoais) [inline phpBB/phpbb/search/fulltext_mysql.php]: @Nicofuma @nickvergessen Updates please. This is still pending...

Comment (nickvergessen) [inline phpBB/phpbb/search/fulltext_mysql.php]: well you can't move `$user` back, so append phpbb_dispatcher to the end so it is optional

Comment (Nicofuma) [inline phpBB/phpbb/search/fulltext_mysql.php]: The best solution is to move everything here to services with a factory but it's out of the scope of  this PR (and can't be done in 3.1 anyway).

About the argument, add it to the end please. Calling a function with too many arguments isn't an issue.

Comment (brunoais) [inline phpBB/phpbb/search/fulltext_mysql.php]: OK... seems reasonable

Comment (brunoais) [inline phpBB/phpbb/search/fulltext_mysql.php]: @nickvergessen @Nicofuma Please confirm if it seems correct.

File: phpBB/includes/acp/acp_main.php (hunk 3 above)
Comment (Nicofuma) [inline phpBB/includes/acp/acp_main.php]: at the end please

File: phpBB/includes/mcp/mcp_post.php [hunk 5]
```diff
@@ -497,7 +497,7 @@ function change_poster(&$post_info, $userdata)
 	{
 		// We do some additional checks in the module to ensure it can actually be utilised
 		$error = false;
-		$search = new $search_type($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $user);
+		$search = new $search_type($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $phpbb_dispatcher, $user);
```
Comment (Nicofuma) [inline phpBB/includes/mcp/mcp_post.php]: at the end please

File: phpBB/install/install_install.php [hunk 6]
```diff
@@ -1492,7 +1492,7 @@ function build_search_index($mode, $sub)
 		set_config_count(null, null, null, $config);
 
 		$error = false;
-		$search = new \phpbb\search\fulltext_native($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $user);
+		$search = new \phpbb\search\fulltext_native($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $phpbb_dispatcher, $user);
```
Comment (Nicofuma) [inline phpBB/install/install_install.php]: at the end please

File: phpBB/install/install_convert.php [hunk 7]
```diff
@@ -796,7 +796,7 @@ function convert_data($sub)
 		}
 
 		$error = false;
-		$convert->fulltext_search = new $search_type($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $user);
+		$convert->fulltext_search = new $search_type($error, $phpbb_root_path, $phpEx, $auth, $config, $db, $phpbb_dispatcher, $user);
```
Comment (Nicofuma) [inline phpBB/install/install_convert.php]: at the end please

File: tests/search/mysql_test.php [hunk 8]
```diff
@@ -37,8 +37,9 @@ protected function setUp()
 		$config['fulltext_mysql_max_word_len'] = 254;
 
 		$this->db = $this->new_dbal();
+		$phpbb_dispatcher = new phpbb_mock_event_dispatcher();
 		$error = null;
 		$class = self::get_search_wrapper('\phpbb\search\fulltext_mysql');
-		$this->search = new $class($error, $phpbb_root_path, $phpEx, null, $config, $this->db, $user);
+		$this->search = new $class($error, $phpbb_root_path, $phpEx, null, $config, $this->db, $phpbb_dispatcher, $user);
```
Comment (Nicofuma) [inline tests/search/mysql_test.php]: at the end please

File: tests/search/native_test.php [hunk 9]
```diff
@@ -33,11 +33,12 @@ protected function setUp()
 		$cache = new phpbb_mock_cache();
 
 		$this->db = $this->new_dbal();
+		$phpbb_dispatcher = new phpbb_mock_event_dispatcher();
 		$error = null;
 		$class = self::get_search_wrapper('\phpbb\search\fulltext_native');
 		$config['fulltext_native_min_chars'] = 2;
 		$config['fulltext_native_max_chars'] = 14;
-		$this->search = new $class($error, $phpbb_root_path, $phpEx, null, $config, $this->db, $user);
+		$this->search = new $class($error, $phpbb_root_path, $phpEx, null, $config, $this->db, $phpbb_dispatcher, $user);
```
Comment (Nicofuma) [inline tests/search/native_test.php]: at the end please

File: tests/search/postgres_test.php [hunk 10]
```diff
@@ -37,8 +37,9 @@ protected function setUp()
 		$config['fulltext_postgres_max_word_len'] = 254;
 
 		$this->db = $this->new_dbal();
+		$phpbb_dispatcher = new phpbb_mock_event_dispatcher();
 		$error = null;
 		$class = self::get_search_wrapper('\phpbb\search\fulltext_postgres');
-		$this->search = new $class($error, $phpbb_root_path, $phpEx, null, $config, $this->db, $user);
+		$this->search = new $class($error, $phpbb_root_path, $phpEx, null, $config, $this->db, $phpbb_dispatcher, $user);
```
Comment (Nicofuma) [inline tests/search/postgres_test.php]: at the end please

**Comment by rxu** [comment]
Also, maybe add typecasting to the constructors arguments?

**Comment by Nicofuma** [comment]
I don't know if it's in the scope of this PR... maybe keep it for when we will put everything in services?

**Comment by rxu** [comment]
Fair enough, agreed.

File: phpBB/includes/mcp/mcp_post.php (hunk 5 above)
Comment (brunoais) [inline phpBB/includes/mcp/mcp_post.php]: Ups

Comment (brunoais) [inline phpBB/includes/mcp/mcp_post.php]: Weird... I see it as being in a different place... I wonder what is wrong...

**Comment by brunoais** [comment]
Typecasting is not in the scope of this PR

**Comment by brunoais** [comment]
!unset WIP

**Comment by Nicofuma** [comment]
and about the version it will be 3.1.6-dev now :/

**Comment by brunoais** [comment]
!unset WIP

File: phpBB/phpbb/search/fulltext_mysql.php [hunk 11]
```diff
@@ -43,6 +43,12 @@ class fulltext_mysql extends \phpbb\search\base
 	protected $db;
 
 	/**
+	 * PhpBB event dispatcher object
```
Comment (marc1706) [inline phpBB/phpbb/search/fulltext_mysql.php]: phpBB

File: phpBB/phpbb/search/fulltext_native.php [hunk 12]
```diff
@@ -87,6 +87,12 @@ class fulltext_native extends \phpbb\search\base
 	protected $db;
 
 	/**
+	 * PhpBB event dispatcher object
```
Comment (marc1706) [inline phpBB/phpbb/search/fulltext_native.php]: phpBB

File: phpBB/phpbb/search/fulltext_postgres.php [hunk 13]
```diff
@@ -56,6 +56,12 @@ class fulltext_postgres extends \phpbb\search\base
 	protected $db;
 
 	/**
+	 * PhpBB event dispatcher object
```
Comment (marc1706) [inline phpBB/phpbb/search/fulltext_postgres.php]: phpBB

File: phpBB/phpbb/search/fulltext_sphinx.php [hunk 14]
```diff
@@ -96,6 +96,12 @@ class fulltext_sphinx
 	protected $dbtype;
 
 	/**
+	 * PhpBB event dispatcher object
```
Comment (marc1706) [inline phpBB/phpbb/search/fulltext_sphinx.php]: phpBB

**Pull request merged**